### PR TITLE
Purchase license route / get permissions route

### PR DESCRIPTION
**Pull request by rschach1**
- first route for purchasing a license implemented following our concept: https://github.com/bettermarks/license-manager-POC/blob/main/docs/concept.md
- also route for 'redeeming' a license (first try)
- already using an external 'hierarchy provider' server

- using now SQLalchemy 2.0 and no longer SQLModel
- Exceptions are raised in the CRUD part. Is this a good idea?

### CODE REVIEW

**Review by cakemanny — COMMENTED** [review]
I left a few comments Rupert.

File: src/licensing/api/api_v1/endpoints/product.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+router = APIRouter()
+
+
+@router.get(
+    "/",
+    response_model=List[product_schema.Product],
+    status_code=http_status.HTTP_200_OK
+)
+async def get_products(session: AsyncSession = Depends(get_async_session)) -> Any:
+    """this is a comment"""
```
[the remark this answers is not in the document]
Comment (rschach1) [reply src/licensing/api/api_v1/endpoints/product.py:20]: Oh :)

File: src/licensing/config.py [hunk 2]
```diff
@@ -3,11 +3,19 @@
 
 
 class Settings(BaseSettings):
-    database_user: str
-    database_password: str
-    database_host: str
-    database_port: str
-    database_name: str
+    # Generic settings
+    API_V1_STR: str = "/v1"
```
[the remark this answers is not in the document]
Comment (rschach1) [reply src/licensing/config.py:7]: yes, right, will remove the 'STR' and replace it by 'PART'

File: src/licensing/hierarchy_provider_client.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+
+async def http_get(url: str, payload: dict | None = None) -> requests.Response:
+    """
+    Sends a GET request to an external API
+    :param url: the url to call
+    :param payload: optional parameters as a dict
+    """
+    try:
+        response = requests.get(
```
[the remark this answers is not in the document]
Comment (rschach1) [reply src/licensing/hierarchy_provider_client.py]: riiight. Done! (hopefully)

File: src/licensing/crud/license.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+from licensing.crud.product import find_product
+from licensing.model import license as license_model
+from licensing.schema import license as license_schema
+
+
+async def check_owners(url: str, purchaser_eid: str, owner_hierarchy_level: str, owner_eids: List[str]) -> bool:
+    """
+    checks, if the requested license owners are in the 'hierarchy path' of the purchaser.
+    That means: is the purchaser 'allowed' to purchase a license for the given owners?
+    This is true, if and only if the purchaser is 'member' of the license owner, f.e.
```
[the remark this answers is not in the document]
Comment (rschach1) [reply src/licensing/crud/license.py:17]: "for example". But it is e.g, right.

File: src/licensing/config.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (rschach1) [reply src/licensing/config.py]: No, it has been removed completely from the settings. I am hardcoding it in the router def ...

File: src/licensing/crud/license.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+    try:
+        hierarchy_list = await get_hierarchy(url, purchaser_eid)
+    except Exception as e:
+        raise HTTPException(
+            status_code=http_status.HTTP_500_INTERNAL_SERVER_ERROR,  # TODO is this the correct error?
+            detail=f"Hierarchy provider server at base URL='{url}' did not respond."
+        )
+
+    # Now do the actual check. TODO Can we do that in a more elegant way?
+    hierarchy_list_as_string = '\t'.join(hierarchy_list)
```
[the remark this answers is not in the document]
Comment (rschach1) [reply src/licensing/crud/license.py]: I completely refactored the result. We no longer do get the hierarchy paths (we do nor not need it), but only the different entities for all 'hierarchy levels'. Something like 
```
[
    "(school)(999)",
    "(school(888)",
    "(class)(344)",
    "(class)(566)",
    "(teacher)(glu::123)"
]
```
will now be returned. That is only the different 'memberships'. We will perform a 'set search', which is O(1) and superfast. I also adapted the route like so:

`GET /hierarchy/users/glu::123/membership`

**Pull request merged**